nx/lib/nx.ex

```diff
@@ -4800,6 +4800,130 @@ defmodule Nx do
+
+  ### Options
+
+    * `:target_rank` - the non-vectorized rank for the output. Can be used to transfer some of the vectorized axes
+    to the non-vectorized shape. If not given, the target rank will be at most the input tensor rank, and can be
+    reduced depending on how many `target_axes` are given.
+
+  ### Examples
+
+      iex> t = Nx.iota({1}, vectorized_axes: [x: 2, y: 3, z: 4])
+      iex> t2 = Nx.revectorize(t, x: 12, y: :auto)
+      iex> t2.vectorized_axes
+      [x: 12, y: 2]
+      iex> t3 = Nx.revectorize(t, a: :auto)
+      iex> t3.vectorized_axes
+      [a: 24]
+
+  Also works on containers. Note that the revectorization happens on a per-entry basis.
+
+      iex> t1  = Nx.iota({1}, vectorized_axes: [x: 2, y: 3])
+      iex> t2 = Nx.iota({1}, vectorized_axes: [x: 2, y: 1])
+      iex> {r1, r2} = Nx.revectorize({t1, t2}, a: :auto)
+      iex> r1.vectorized_axes
+      [a: 6]
+      iex> r2.vectorized_axes
+      [a: 2]
+
+  This function is useful for when you need to introduce a temporary custom axis to ease calculations.
+  The example below shows how to manipulate your vectorized tensor for that objective.
+
+      iex> t = Nx.iota({2}, vectorized_axes: [x: 2, y: 2])
+      #Nx.Tensor<
+        vectorized[x: 2][y: 2]
+        s64[2]
+        [
+          [
+            [0, 1],
+            [0, 1]
+          ],
+          [
+            [0, 1],
+            [0, 1]
+          ]
+        ]
+      >
+      iex> revec = Nx.revectorize(t, temp: :auto)
+      #Nx.Tensor<
+        vectorized[temp: 4]
+        s64[2]
+        [
+          [0, 1],
+          [0, 1],
+          [0, 1],
+          [0, 1]
+        ]
+      >
+      iex> result = Nx.add(revec |> Nx.vectorize(x: 2), Nx.tensor([1, -1]) |> Nx.vectorize(x: 2))
+      iex> Nx.revectorize(result, t.vectorized_axes, target_rank: Nx.rank(t))
+      #Nx.Tensor<
+        vectorized[x: 2][y: 2]
+        s64[2]
+        [
+          [
+            [1, 0],
+            [1, 0]
+          ],
+          [
+            [1, 0],
+            [1, 0]
+          ]
+        ]
+      >
+
+  Note how we could reuse the `:x` name because through `revectorize/3` we ensured
+  that it was no longer a vectorized axis name. This is especially important if we
+  are receiving unknown tensors, as this enables us to force specific names for
+  internal calculations, and then we can restore the original ones for the output.
+  """
+  @doc type: :vectorization
+  def revectorize(tensor, target_axes, opts \\ [])
+
+  def revectorize(%T{} = tensor, target_axes, opts) do
+    opts = keyword!(opts, [:target_rank])
+    auto_axes = Enum.count(target_axes, fn {_k, v} -> v == :auto end)
+
+    if auto_axes > 1 do
+      raise ArgumentError,
+            "at most one vectorized axis size can be calculated automatically, got: #{auto_axes}"
+    end
+
+    axes_names = Keyword.keys(target_axes)
+
+    {names, devectorized_shape_l} =
+      if target_rank = opts[:target_rank] do
```

Comment:
What happens if rank is more than the current tensor rank?

Reply:
Then the last vectorized axes will be absorbed into the tensor rank.
The idea of this option is to enable a way for this process to happen.
Maybe we can rework the option into `:pop` or something like this, being the number of trailing vectorized axes that will be absorbed into the inner shape.

Though the current interface is a bit easier because it allows us to restore the original tensor shape and axes by passing these fields as the `target_axes` and `:target_rank`

Reply:
sorry, i mean that if you say target_rank 5 and the current devectorized tensor  has rank 2, we have to raise, no?

Reply:
ah, yes